### PR TITLE
snippets pkg: MarshalJSON and DeepCopy

### PR DESCRIPTION
**Pull request by yuzhouliu9**
This is the first submission for components of the `snippets` pkg:

- [x] MarshalJSON
- [x] DeepCopy, DeepCopyInto
- [ ] mergeMapStringString
- [ ] Constructor
- [ ] Default Builder methods (With...)

The functions will take in the `context`, `type`, and write out the generated code to the `io.Writer`.
For GenerateMarshalJSON, it also requires an additional input argument for `embeddedTypePackage` name.

### CODE REVIEW

File: pkg/generators/snippets/marshaljson.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"errors"
+	"io"
+
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+)
+
+func GenerateMarshalJSON(ctx *generator.Context, t *types.Type, embeddedTypePackage string, w io.Writer) error {
```
Comment (yuzhouliu9) [inline pkg/generators/snippets/marshaljson.go:11]: I chose to have `embeddedTypePackage string` as an input parameter, and leave the finding and importing of the embedded type to the calling code.

The other way is to have imports as an input parameter. In that case, the calling code would need to `imports.AddType(embeddedType)`, and within this function we'd need to find the embeddedType from the  `t` type, and then do `"alias": imports.LocalNameOf(embeddedType.Name.Package)` on L18.

I think the current way is cleaner but open to suggestions.

Comment (dskatz) [reply pkg/generators/snippets/marshaljson.go:11]: I that this is fine.  These functions should only implement the snippet writer.

My suggestion is finding a better name for `embeddedTypePackage`.   Maybe, `schemaGroupVersionPackageAlias`?

**Review by dskatz — COMMENTED** [review]
Looks good.  

My one other question is should the signature of these functions follow the pattern of `NewSnippetWriter`?

`generator.NewSnippetWriter(w, ctx, `

e.g.
```
GenerateDeepCopy(w io.Writer, ctx *generator.Context, t *types.Type)
```

**Comment by yuzhouliu9** [comment]
> Looks good.
> 
> My one other question is should the signature of these functions follow the pattern of `NewSnippetWriter`?
> 
> `generator.NewSnippetWriter(w, ctx, `
> 
> e.g.
> 
> ```
> GenerateDeepCopy(w io.Writer, ctx *generator.Context, t *types.Type)
> ```

Yup, thanks. I reordered the parameter list to follow pattern of `NewSnippetWriter`

**Review by dskatz — APPROVED** [review]
LGTM. I think as we build out the generator I can see adding integration tests for these snippets to evaluate compilation.

**Pull request merged**